source/extensions/filters/http/stateful_session/stateful_session.cc

```diff
@@ -0,0 +1,84 @@
+#include "source/extensions/filters/http/stateful_session/stateful_session.h"
+
+#include <cstdint>
+#include <memory>
+
+#include "source/common/config/utility.h"
+#include "source/common/http/utility.h"
+#include "source/common/upstream/load_balancer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace StatefulSession {
+
+StatefulSessionConfig::StatefulSessionConfig(const ProtoConfig& config,
+                                             Server::Configuration::CommonFactoryContext& context) {
+  if (config.host_statuses_size() > 0) {
+    std::vector<envoy::config::core::v3::HealthStatus> host_statuses;
+    host_statuses.reserve(config.host_statuses_size());
+    for (int i = 0; i < config.host_statuses_size(); i++) {
+      host_statuses.push_back(config.host_statuses(i));
+    }
+    host_statuses_ = Upstream::LoadBalancerContextBase::createOverrideHostStatus(host_statuses);
+  } else {
+    host_statuses_ = ~host_statuses_;
+  }
+
+  auto& factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<Envoy::Http::SessionStateFactoryConfig>(
+          config.session_state().name());
+
+  auto typed_config = Envoy::Config::Utility::translateAnyToFactoryConfig(
+      config.session_state().typed_config(), context.messageValidationVisitor(), factory);
+
+  factory_ = factory.createSessionStateFactory(*typed_config, context);
+}
+
+PerRouteStatefulSession::PerRouteStatefulSession(
+    const PerRouteProtoConfig& config, Server::Configuration::CommonFactoryContext& context) {
+  if (config.override_case() == PerRouteProtoConfig::kDisabled) {
+    disabled_ = true;
+    return;
+  }
+  config_ = std::make_shared<StatefulSessionConfig>(config.stateful_session(), context);
+}
+
+Http::FilterHeadersStatus StatefulSession::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  const StatefulSessionConfig* config = config_;
+  auto route_config = Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteStatefulSession>(
+      "envoy.filters.http.stateful_session", decoder_callbacks_->route());
+
+  if (route_config != nullptr) {
+    if (route_config->disabled()) {
+      return Http::FilterHeadersStatus::Continue;
+    }
+    config = route_config->statefuleSessionConfig();
```

Comment:
Are we overriding the initialized config here? If it is per filter state, should we have it in this filter class?

Reply:
I will check it.

Reply:
> Are we overriding the initialized config here?

Yes. The route level filter config will override the listener level filter config.

> If it is per filter state, should we have it in this filter class?

Currently it is not needed, because we only use it in the decode phase.